### PR TITLE
Issue #5325: MySQL on PHP 8 now errors when committing or rolling back when there is no active transaction.

### LINKED ISSUE #5325: Fixed: PHP 8: Database: Transaction test failure
The database transaction test fails on PHP 8 with the following output:

```
---- Database: Transaction tests (DatabaseTransactionTestCase) ----


Status    Group      Filename          Line    Function                            
------------------------------------------------------------------------------------------------------------------------
Exception Uncaught e database.inc      285     commit()                                                                
    PDOException: There is no active transaction in BackdropTestCase->run() (line 617 of
    /home/runner/work/backdrop/backdrop/core/modules/simpletest/backdrop_web_test_case.php).
Error: Process completed with exit code 1.
```

This matches https://www.drupal.org/project/drupal/issues/3204161, which has been fixed in https://git.drupalcode.org/project/drupal/-/commit/0f08b88

Part of meta PHP 8 issue at https://github.com/backdrop/backdrop-issues/issues/5076

### PR DESCRIPTION
**Pull request by quicksketch**
Fixes https://github.com/backdrop/backdrop-issues/issues/5325.

Port of [Drupal issue #3204161](https://www.drupal.org/project/drupal/issues/3204161) by mcdruid, alexpott, catch, Charlie ChX Negyesi, mondrake, Mile23, andypost.

### CODE REVIEW

File: core/includes/database/mysql/database.inc [hunk 1]
```diff
@@ -315,6 +315,53 @@ class DatabaseConnection_mysql extends DatabaseConnection {
...
+      // In PHP 8.0 a PDOException is thrown when a commit is attempted with no
+      // transaction active. In previous PHP versions this failed silently.
+      return TRUE;
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function rollback($savepoint_name = 'drupal_transaction') {
```
Comment (indigoxela) [inline core/includes/database/mysql/database.inc]: Cough... *backdrop*_transaction would be better here :wink:

File: core/modules/simpletest/tests/database_test.test [hunk 2]
```diff
@@ -3504,21 +3504,36 @@ class DatabaseTransactionTestCase extends DatabaseTestCase {
...
 
+  /**
+   * Special handling of "rollback without transaction" errors.
+   */
+  public function rollBackWithoutTransactionErrorHandler($error_level, $message, $filename, $line) {
+    // Throw an exception if this is a "rollback without transaction" error.
+    if (strpos($message, 'Rollback attempted when there is no active transaction.') !== FALSE ) {
+      throw new Exception('Rollback attempted when there is no active transaction.');
+    }
+    _drupal_error_handler($error_level, $message, $filename, $line);
```
Comment (indigoxela) [inline core/modules/simpletest/tests/database_test.test]: Similar as above _**backdrop**_error_handler, function _drupal_error_handler doesn't exist in Backdrop

**Review by klonos — APPROVED** [review]
LGTM 👍🏼

**Comment by quicksketch** [comment]
Thanks @klonos and @indigoxela! I merged this into 1.x and 1.20.x.

**Pull request merged**